File: mealymachine.py
class Utility:
    @staticmethod
    def show_error(message):
        print("Output: Your string is not valid")
        print("Reason: " + message)

    @staticmethod
    def get_input(message):
        while True:
            input_string = input(message)
            if len(input_string) == 0:
                print("Input can not be empty")
            else:
                return input_string


class MealyMachineValidateString:
    @staticmethod
    def main():
        message = "Enter valid input symbols (comma-separated): "
        valid_input_string = Utility.get_input(message)
        valid_inputs = set(valid_input_string.split(","))

        message = "Enter states (Q) (comma-separated): "
        state_string = Utility.get_input(message)
        states = set(state_string.split(","))

        message = "Enter starting state (q0): "
        starting_state = Utility.get_input(message)

        transitions = {}

        for state in states:
            transitions[state] = {}
            for valid_input in valid_inputs:
                message = f"Enter next state for transition '{state}' with input '{valid_input}': "
                next_state = Utility.get_input(message)
                message = f"Enter output for transition '{state}' with input '{valid_input}': "
                output = Utility.get_input(message)
                transitions[state][valid_input] = (next_state, output)

        mealy_machine = MealyMachine(starting_state, transitions)

        while True:
            message = "Enter input sequence to be processed: "
            input_sequence = Utility.get_input(message)
            mealy_machine.process_input_sequence(input_sequence)

            choice = input("Do you want to process another input sequence? (y/n): ")
            if choice.lower() == "n":
                break


class MealyMachine:
    def __init__(self, starting_state, transitions):
        self.starting_state = starting_state
        self.transitions = transitions

    def process_input_sequence(self, input_sequence):
        current_state = self.starting_state
        output_sequence = ""
        traversal_path = [current_state]

        for input_symbol in input_sequence:
            current_state_transition = self.transitions.get(current_state, {})
            if input_symbol in current_state_transition:
                next_state, output = current_state_transition[input_symbol]
                output_sequence += output
                traversal_path.append(f"{current_state}({input_symbol},{output})->{next_state}")
                current_state = next_state
            else:
                Utility.show_error(f"Invalid transition: State '{current_state}' with input '{input_symbol}'")
                return

        print("Output Sequence:", output_sequence)
        print("Traversal Path:", " -> ".join(traversal_path))


if __name__ == "__main__":
    MealyMachineValidateString.main()

File: mealymachine.java
import java.util.*;

class Main {
    public static void main(String[] args) {
        MealyMachineValidateString.main();
    }
}

class Utility {
    
    public static String get_input(String message) {
        Scanner scanner = new Scanner(System.in);
        String inputString;

        do {
            System.out.print(message);
            inputString = scanner.nextLine();
            if (inputString.length() == 0) {
                System.out.println("Input can not be empty");
            }
        } while (inputString.length() == 0);

        return inputString;
    }
}

class MealyMachineValidateString {
    public static void main() {
        String inputString, message;

        message = "Enter valid input symbols (comma-separated): ";
        inputString = Utility.get_input(message);
        Set<String> validInputs = new HashSet<>(Arrays.asList(inputString.split(",")));

        message = "Enter states (Q) (comma-separated): ";
        inputString = Utility.get_input(message);
        Set<String> states = new HashSet<>(Arrays.asList(inputString.split(",")));

        message = "Enter starting state: ";
        String startingState = Utility.get_input(message);

        Map<String, Map<String, String>> transitions = new HashMap<>();

        for (String state : states) {
            Map<String, String> transition = new HashMap<>();
            for (String validInput : validInputs) {
                message = String.format("Enter next state and output for transition '%s' with input '%s': ", state, validInput);
                String next_state_and_output = Utility.get_input(message);
                transition.put(validInput, next_state_and_output);
            }
            transitions.put(state, transition);
        }

        MealyMachine mealyMachine = new MealyMachine(startingState, transitions);

        while (true) {
            message = "Enter input sequence to be processed: ";
            String input_sequence = Utility.get_input(message);
            mealyMachine.process_input_sequence(input_sequence);

            System.out.print("Do you want to process another input sequence? (y/n): ");
            String choice = new Scanner(System.in).nextLine();
            if (choice.equalsIgnoreCase("n")) {
                break;
            }
        }
    }
}

class MealyMachine {
    private String startingState;
    private Map<String, Map<String, String>> transitions;

    public MealyMachine(String startingState, Map<String, Map<String, String>> transitions) {
        this.startingState = startingState;
        this.transitions = transitions;
    }

    public void process_input_sequence(String input_sequence) {
        String current_state = startingState;
        String output_sequence = "";
        List<String> traversal_path = new ArrayList<>();
        traversal_path.add(current_state);

        for (int i = 0; i < input_sequence.length(); i++) {
            String input_symbol = String.valueOf(input_sequence.charAt(i));
            Map<String, String> current_transitions = transitions.get(current_state);

            if (current_transitions.containsKey(input_symbol)) {
                String next_state_and_output = current_transitions.get(input_symbol);
                String[] next_state_and_output_arr = next_state_and_output.split(",");
                String next_state = next_state_and_output_arr[0];
                String output = next_state_and_output_arr[1];

                output_sequence += output;
                traversal_path.add(current_state + "(" + input_symbol + "," + output + ")->" + next_state);
                current_state = next_state;
            } 
        }

        System.out.println("Output Sequence: " + output_sequence);
        System.out.println();
        System.out.println("Traversal Path: " + String.join(" -> ", traversal_path));
    }

    
}
